nautobot/dcim/models/devices.py

```diff
@@ -979,3 +1034,122 @@ def devices_sorted(self):
 
     def __str__(self):
         return self.name
+
+
+#
+# Software images
+#
+
+
+class SoftwareImageQuerySet(RestrictedQuerySet):
+    """Queryset for SoftwareImage objects."""
+
+    def get_for_object(self, obj):
+        """Return all SoftwareImage assigned to the given object."""
+        if not isinstance(obj, models.Model):
+            raise TypeError(f"{obj} is not an instance of Django Model class")
+        if isinstance(obj, Device):
+            qs = self.filter(software_version__platform__devices=obj)
+        elif isinstance(obj, InventoryItem):
+            qs = self.filter(software_version__platform__devices__inventory_items=obj)
+        else:
+            qs = self
+
+        return qs
+
+
+@extras_features(
+    "custom_links",
+    "custom_validators",
+    "export_templates",
+    "graphql",
+    "statuses",
+    "webhooks",
+)
+class SoftwareImage(PrimaryModel):
+    """A software image for a Device, Virtual Machine or Inventory Item."""
+
+    software_version = models.ForeignKey(
+        to="SoftwareVersion", on_delete=models.CASCADE, related_name="software_images", verbose_name="Software Version"
+    )
+    image_file_name = models.CharField(blank=False, max_length=255, verbose_name="Image File Name")
+    image_file_checksum = models.CharField(blank=True, max_length=256, verbose_name="Image File Checksum")
+    hashing_algorithm = models.CharField(
+        choices=SoftwareImageHashingAlgorithmChoices,
+        blank=True,
+        max_length=255,
+        verbose_name="Hashing Algorithm",
+        help_text="Hashing algorithm for image file checksum",
+    )
+    image_file_size = models.PositiveBigIntegerField(
+        blank=True,
+        null=True,
+        verbose_name="Image File Size",
+        help_text="Image file size in bytes",
+    )
+    download_url = models.URLField(blank=True, verbose_name="Download URL")
+    status = StatusField(blank=False, null=False)
+
+    objects = BaseManager.from_queryset(SoftwareImageQuerySet)()
+
+    class Meta:
+        ordering = ("software_version", "image_file_name")
+        unique_together = ("image_file_name", "software_version")
+
+    def __str__(self):
+        return self.image_file_name
```

Comment:
Should we include `software_version` here as well since `image_file_name` is not unique by itself?